### ISSUE TITLE
Message subscription and dispatch

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/trin#102]
> My model of rust's execution model is minimal at best. Any help in mapping this into rust conventions is appreciated.

In the current code we spawn a background process `process_discv5_requests` that handles incoming requests:

https://github.com/ethereum/trin/blob/5f75c0532a1c8ba59589d3482ea66a109ece1c4c/trin-history/src/lib.rs#L68-L70

The `process_discv5_requests` enters an infinite loop, picking up one inbound packet, checking that it is a `TalkRequest` and then handling it in `process_one_request` which returns a response message which is then dispatched back to the original sender.

The handling process is implemented with a `match` against each of the different message types, constructing the response packet and returning it.

This architecture is unlikely to facilitate a clean and rubust client for multiple reasons.

## Doing *work* after handling a request

We need the ability to do *work* after dispatching the response packet.  An example:

The OFFER/ACCEPT message pair are used to initiate a uTP stream over which the actual data will be transferred.  For this to work, the node receiving the OFFER message will send back an ACCEPT message with a `connection-id` for the uTP stream.  After sending the ACCEPT back the receiving node will need to listen for an incoming uTP stream with that connection-id.  There is likely additional validation needed after receiving the whole payload.  Timeouts will also be a thing during this sequence of messages.

If the handling of the message happens within the main message handling loop, then we will end up blocking the processing of messages until this sequence has completed.

## Doing *maintenance* work.

Every PONG message containes an ENR sequence number.  Every time we receive a pong, we should check whether the sequence number is newer than our local record.  In the event of a higher sequence number we need to request an updated ENR record from the node.

PING messages are going to be used in a number of places.  One specific use is validating liviliness which happens in a few different contexts.  We **do not** want to copy/paste this sequence number checking code across these places.  We also **do not** want to have to *remember* to trigger this process each time we use the PING message.

The only place where PONG messages can currently be handled in the current architecture is at the call site where the ping was sent, which is very limiting.

The solution to this is to implement an API that allows *subscriptions* to specific message types.  This allows us to spawn a background process which subscribes and is given a chance to process every inbound PONG message.

This same subscription API can also easily be used to implement standard handling of request messages.  Here is the handling of inbound PING requests in the python implementation of this:

https://github.com/ethereum/ddht/blob/142911d134ff839f3f79ff8fe9e45d3fe5a58cd0/ddht/v5_1/alexandria/network.py#L290-L308

## The suggested architecture

Starting with the main stream of `TALKREQ` and `TALKRESP` messages.

```python
def process_requests(...):
    for message in stream_of_messages:
        subscription_manager.handle_message(message)
```

The `subscription_manager` in this case is doing something like this:

```
class SubscriptionManager:
    def handle_message(message):
        subscriptions = self.get_subscriptions_for_message_type(type(message))
        for subscription in subscriptions:
            # the message just gets put in a queue to be consumed elsewhere
            subscription.append(message)
```

What this facilitates is the ability to implement message handling as long running background processes, instead of doing it in the main handling loop.  This ensures that the main handling loop isn't blocked by messages that take a long time to process.

**Comment by pipermerriam** [comment]
It looks like `tokio` has tools to do this here: https://docs.rs/tokio/0.2.11/tokio/sync/broadcast/fn.channel.html

**Comment by victor-wei126** [comment]
> If the handling of the message happens within the main message handling loop, then we will end up blocking the processing of messages until this sequence has completed.

By "main message handling loop" do you mean the loop in [process_discv5_requests](https://github.com/ethereum/trin/blob/master/trin-core/src/portalnet/protocol.rs#L202-L223)? If so, I am confused on how doing work after handling a request will block the loop, since I it seems to me as if any work will be done outside of the loop, and thus outside of the thread since the loop is the only thing being run in the thread. For example, in the OFFER/ACCEPT message pair there is a function `send_offer`, which then waits for the ACCEPT message. Once OFFER is sent, the handling is done in the `process_one_request` loop, which sends back ACCEPT. Now we are back in the `send_offer` method, and here we can have code to open a utp stream and pipe the data over. This data will be handled by the receiving node, but it seems that all of this work will be done outside of the loop we currently have, so it won't block. But my knowledge of threads is limited.

> The only place where PONG messages can currently be handled in the current architecture is at the call site where the ping was sent, which is very limiting.

On the topic of doing maintenance work, I suppose my previous point might be moot cuz even if it is not blocking, the handling of ACCEPT can only be done at the spot where OFFER originated, which may be limiting.

I also remember we talked about the fact that the `process_discv5_requests` is spawned on a background process, and the code is blocking since when we are handling (matching) the request, we can only handle this one request at this time, and can't handle other incoming requests, iirc. But is this a big issue? since I imagine the handling to be relatively quick. maybe the implementors of this can give more insight on the design?

**Comment by SamWilsn** [comment]
While I don't think you're implying using a single `broadcast` channel for all message types @pipermerriam, I want to point out that these channels will drop messages on slow receivers:

> If the Receiver handle falls behind, once the channel is full, newly sent values will overwrite old values.

---

I am _far_ from a tokio expert and I haven't been keeping up with trin as much as I should be, so take everything I say with a grain of salt.

One super basic "solution" is to just `spawn` every task that handles a message from the event loop and let tokio's scheduler sort out the priority. This would allow I/O heavy operations to complete while keeping the even loop responsive, but wouldn't really help much with CPU bound tasks (as far as I understand concurrency vs. parallelism...)

This approach is nice because it really doesn't change much from the basic event loop paradigm:

```
def event_loop():
    while True:
        msg = receive()
        if msg is PONG:
            handle_pong(msg)
        elif msg is ...:
            ....

def handle_pong(msg):
    # TODO: validation
    spawn(update_table(msg))
    spawn(update_liveness(msg))
```

There are several downsides to this approach:

 - Debugging sucks: little visibility into where in the system a message is.
 - Resource management sucks: doing rate limiting, tracking, or any other kind of accounting is difficult since all you have are tasks in the executor.
 - No formal structure: you have to trace every message through several layers of functions to see the order of assertions.
 - No easy state tracking: original message context is lost.

---

Whatever approach we end up going with, I think I'd like to support a couple usage patterns:

 1. Command/response - the task which sends a message continues sequentially after receiving a response:
     ```
     let foo = bar.send().await;
     foo.do_something_with_the_response();
     ```
 1. Subscriptions/validation - code that runs for a message kind regardless of who sent it (likely before the reply handler is invoked.)

I don't think we necessarily need to go with a broadcast/subscriber model, especially since I don't think the handlers are dynamic, but something more formal than the event loop might be nice too.

**Comment by ogenev** [comment]
> One super basic "solution" is to just `spawn` every task that handles a message from the event loop and let tokio's scheduler sort out the priority. This would allow I/O heavy operations to complete while keeping the event loop responsive, but wouldn't really help much with CPU bound tasks (as far as I understand concurrency vs. parallelism...)

I'm also leaning towards the "event loop" approach. I like it because it is simple and keep the added complexity low. Iterating on it, probably it makes sense to spawn one main thread for every handler, and then, if needed, we can spawn additional background threads. So it may looking something like:

```python
def event_loop():
    while True:
        msg = receive()
        if msg is PONG:
            spawn(handle_pong(msg))
        elif msg is ...:
            ....

def handle_pong(msg):
    # TODO: validation
    # spawn background threads if needed
    spawn(update_table(msg))
    spawn(update_liveness(msg))
```

My concern about the single channel "broadcast" approach is that sending values faster than they can be received and processed will cause an ever-growing backlog of values to accumulate in the channel.

**Comment by pipermerriam** [comment]
I think our use case is going to be tolerant of dropping messages when the subscription channel is not processing them fast enough.  The spawn approach to message handling sounds fine to me.  I do still think there is a need for a subscription approach but I support us making the minimal change necessary to support the functionality that we need now.

**Comment by victor-wei126** [comment]
So if we are going with the event loop approach for now, we already have an event loop `process_discv5_requests()` which processes requests in a loop. This processes any portalnetwork requests, as well as utp requests, but it also processes utp byte streams which would raise the blocking issue (but also as a side issue this is not a request so it doesn't seem to fit) but we could solve it by spawning on a new task? Would we have a separate loop which processes the responses then, but is otherwise analogous?

**Comment by lithp** [comment]
I'm coming to this issue seeing only what was discussed here so forgive me if a decision was reached elsewhere and I'm commenting on a stale issue.

To rephrase what I think was discussed, currently incoming messages are processed by a single thread with a single flow of control. A single-threaded eventloop is easy* to write and easy to debug, but any work done in that thread blocks the processing of future messages. I'll add that blocking might not be so bad! Even if it takes a snail's 100ms to handle each message that still works out to 10 handled messages per second which seems like more than enough for a testnet.

I said using an eventloop is easy* because it's easy as long as the protocol is stateless, meaning e.g. you can handle a PING by doing some processing and sending a PONG and then forgetting all about the PING. It becomes more complicated as you need to remember details of previous packets, such as needing to remember the set of `connection-id`s which have been passed out. Correctly handling timeouts by remembering when each of those `connection-ids` expires adds another level of complication.

I don't know enough yet about what work needs to be done to judge whether an eventloop would be too complicated, but just going off of what is in this thread I bet the loop will be relatively clean and run plenty fast for our purposes, there's not a lot of state manipulation which needs to happen.

---

Option 2 is to use a more modern eventloop w/ coroutines. At the end of the day there's still a single thread accepting messages and deciding what to do with them, so the performance isn't any better, but the loop is abstracted away from most code. Instead, you get to write e.g. the multi-step OFFER/ACCEPT flow as if it was the only thing happening on the thread, blissfully ignorant of the other tasks running concurrently.

As Sam mentioned this can make debugging (especially printf debugging) and other forms of visibility harder. At least, it does in most languages, I don't know what the Rust tooling is like. If we have a lot of stateful interactions then option 2 will be simpler even though it makes debugging harder.

---

I don't think I understand the subscription-based approach, it moves handling of e.g. PING requests to a background thread but the listener is already a background thread. It prevents PING message handling from blocking PONG message handling (though, in the linked python version they're still running in the same thread, and block each other!) but PING messages can still block each other.

---

I wanted to end with a proposed architecture, I think it's possible to get something fairly clean by doing all the maintenance work in-line with the eventloop but providing enough hooks that the consumer of the Overlay API can do the command/response thing Sam mentioned. It's getting late here though so I think I'll end here and try to write a demonstration soon.

**Comment by victor-wei126** [comment]
Hi! Nope this is still a very much active issue and I've been trying to come up with a working solution for it, but would be glad to have your help/insights.

With regards to your comments on the current single-threaded eventloop solution, yes I agree, the blocking isn't bad at all since the subnetwork event handlers currently just proxy incoming requests to the overlay protocol `process_one_request()` method, which is really just a large if/else statement, so it shouldn't be fast to process at all.

> It becomes more complicated as you need to remember details of previous packets, such as needing to remember the set of connection-ids which have been passed out. Correctly handling timeouts by remembering when each of those connection-ids expires adds another level of complication.

I am not sure what you mean by this, but afaik the `connection-ids` are only used in the utp streams, which only will be initiated in the OFFER/ACCEPT or FINDCONTENT messages. I'm not so sure about the FINDCONTENT use case, but the node sending ACCEPT will listen for an incoming utp connection on the given `connection-id` it sends out. The node that receives ACCEPT is the one that is actually responsible for initiating the connection and stream. I believe this is handled by utp, since a new `connection` will be added to the `utp_connections` hash map. But now that you bring it up, yes, the node waiting for a connection will need to maintain some sort of state, but I believe this is the only state management necessary as of now.

> Option 2 is to use a more modern eventloop w/ coroutines. At the end of the day there's still a single thread accepting messages and deciding what to do with them, so the performance isn't any better, but the loop is abstracted away from most code. Instead, you get to write e.g. the multi-step OFFER/ACCEPT flow as if it was the only thing happening on the thread, blissfully ignorant of the other tasks running concurrently.

I was under the impression that since we are spawning additional threads for additional handling of each message, this avoids any blocking issues and allows for better performance. What do you mean by "abstracting the loop away from the code"? I would be interested in seeing how the OFFER/ACCEPT flow be implemented in your design, as I am working on an implementation of that now.

**Comment by victor-wei126** [comment]
Also, here is my stab at an architecture. Currently, the subnetwork (state, history) specific event handlers receive network specific events on the receiver end of a channel. They are waiting for incoming requests in a loop, and once a request comes in, the handler immediately sends the message to the overlay protocol to handle it, in the `process_one_request()` method. Within that method, all that is being done is matching the request to the appropriate response, then building a response message and sending it back to the subnetwork event handler. The subnet event handler then responds to the request with the reply, and I believe this is done directly with discv5. So this whole process successfully receives requests from discv5, does the appropriate handling, and replies with the correct response.

Now comes the issue of how does the node _receiving_ the response handle the response. Well, the node receiving the response must have issued the request in the first place, and this is done in the `send_*` messages, where * is one of the request messages (PING, OFFER, etc). I am not sure where these messages are called, but what I do know is that they return an array of bytes, ie, the response message, directly from discv5. I think this was the issue Piper was talking about in the original post, where "the only places where responses can be handled is at the call site where the request was sent".

So putting this all together. In the proposed eventloop architecture, we have a line `msg = receive()`. This seems to me where we would call something like `let msg = send_*().await`. Then, we match the `msg` to a response, and then handle the response appropriately, whether that be spawning a new thread. So something like this:

```
pub async fn process_responses(&self) {
        loop {
            let response = send_ping().await;
            let response = send_find_content().await;
            ....

            match response {
                Response::Pong(Pong {..}) => {
                    // handle(PONG)
                },
                Response::Nodes(Nodes {..}) => {
                    // handle(NODES)
                },
                Response::Accept(Accept {..}) => {
                    // handle(ACCEPT)
                },
            }
        }
    }
```
The first chunk of that code doesn't make sense however, since it just sends out a request, gets the response. But I hope it illustrates the point that currently the `send_*()` methods are receiving the responses and we need to call await on those methods to get the response back. My only thought at this point is to create a channel. Within the `send_*()` methods, instead of returning the response directly, it instead uses a transmitter to send the bytes down a channel, which is received by the receiver. We then have a loop listening for events received on the receiver. Any feedback on this would be appreciated.
